### PR TITLE
Implement syncbn for TensorFlow

### PR DESCRIPTION
**Pull request by edwardyehuang**
Add SyncBN implementation (via `synchronized` argument in `layers.BatchNormalization`) for TensorFlow. Note that, I don't know how to write the test for it in Keras online test (it requires multi-gpus).

#18667

### CODE REVIEW

**Comment by edwardyehuang** [comment]
Also, I am not sure if it is better to move the framework-specified code (e.g. TensorFlow) to `backbend`, the current `distribution_lib.py` states "!!!DO NOT USE!!!"

File: keras/layers/normalization/batch_normalization.py [hunk 1]
```diff
@@ -67,6 +68,10 @@ class BatchNormalization(Layer):
         gamma_regularizer: Optional regularizer for the gamma weight.
         beta_constraint: Optional constraint for the beta weight.
         gamma_constraint: Optional constraint for the gamma weight.
+        synchronized: If True, synchronizes the global batch statistics (mean and
+        variance) for the layer across all devices at each training step in a
```
Comment (fchollet) [inline keras/layers/normalization/batch_normalization.py:72]: Add indent (4 spaces)

File: keras/layers/normalization/batch_normalization.py [hunk 2]
```diff
@@ -67,6 +68,10 @@ class BatchNormalization(Layer):
         gamma_regularizer: Optional regularizer for the gamma weight.
         beta_constraint: Optional constraint for the beta weight.
         gamma_constraint: Optional constraint for the gamma weight.
+        synchronized: If True, synchronizes the global batch statistics (mean and
```
Comment (fchollet) [inline keras/layers/normalization/batch_normalization.py:71]: Use backticks around code keywords like True/False

File: keras/layers/normalization/batch_normalization.py [hunk 3]
```diff
@@ -125,10 +130,15 @@ def __init__(
...
         beta_constraint=None,
         gamma_constraint=None,
+        synchronized=False,
         **kwargs,
     ):
         super().__init__(**kwargs)
         self.axis = int(axis)
+
+        self.synchronized = synchronized
+        # TODO: Set fused=False if synchronized is True
```
Comment (fchollet) [inline keras/layers/normalization/batch_normalization.py:140]: We don't have fused BN so you can remove this

File: keras/layers/normalization/batch_normalization.py [hunk 4]
```diff
@@ -187,6 +197,65 @@ def build(self, input_shape):
 
     def compute_output_shape(self, input_shape):
         return input_shape
+    
+
+    def calculate_mean_and_var(self, inputs):
```
Comment (fchollet) [inline keras/layers/normalization/batch_normalization.py:202]: Make these 2 methods private

File: keras/layers/normalization/batch_normalization.py [hunk 5]
```diff
@@ -187,6 +197,65 @@ def build(self, input_shape):
...
+
+        if not self.synchronized:
+            return ops.moments(
+                inputs, axes=self._reduction_axes, keepdims=True
+            )
+        else:
+            return self.calulate_sync_mean_and_var(inputs)
+
+
+    def calulate_sync_mean_and_var(self, inputs):
```
Comment (fchollet) [inline keras/layers/normalization/batch_normalization.py:212]: Since it's only used once you can merge it with the method above (you don't even need the indented `else:` since the first branch returns)

File: keras/layers/normalization/batch_normalization.py [hunk 6]
```diff
@@ -187,6 +197,65 @@ def build(self, input_shape):
...
+
+        axes = self._reduction_axes
+        y = inputs
+
+        if backend() == "tensorflow":
+            
+            from backend.tensorflow.distribution_lib import tf
```
Comment (fchollet) [inline keras/layers/normalization/batch_normalization.py:219]: Do: `from keras.utils.module_utils import tensorflow as tf`

File: keras/layers/normalization/batch_normalization.py [hunk 7]
```diff
@@ -125,10 +130,15 @@ def __init__(
         gamma_regularizer=None,
         beta_constraint=None,
         gamma_constraint=None,
+        synchronized=False,
         **kwargs,
     ):
         super().__init__(**kwargs)
         self.axis = int(axis)
+
+        self.synchronized = synchronized
```
Comment (fchollet) [inline keras/layers/normalization/batch_normalization.py:139]: Add: 

```
if synchronized and backend.backend() != "tensorflow":
    raise ValueError("Argument synchronized=True is only supported with the TensorFlow backend.")
```
```

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

@qlzh727 can you advise on how to proceed for testing the feature with 2 devices?

**Review by fchollet — APPROVED** [review]
LGTM, thanks. Will post-process this on our side.

**Comment by qlzh727** [comment]
For sync batch norm test, user can config 2 virtual cpus and use them for mirrored strategy test.

https://www.tensorflow.org/api_docs/python/tf/config/set_logical_device_configuration

**Comment by qlzh727** [comment]
ack, I will add a test for that. The sync logic will only have a difference in the distribution setting.
